added PredisProvider

ivoba opened the pull request:
added PredisProvider
added functional test
improved options configuration
formatting of riak.xml in tests (off topic)

ivoba commented:
relates to https://github.com/doctrine/DoctrineCacheBundle/issues/44

guilhermeblanco commented on composer.json:
```diff
@@ -49,7 +49,8 @@
         "instaclick/object-calisthenics-sniffs": "dev-master",
-        "instaclick/symfony2-coding-standard":   "dev-remaster"
+        "instaclick/symfony2-coding-standard":   "dev-remaster",
+        "predis/predis":                         "~0.8"
```
Make this part of suggest section, not a dev requirement.
If I'm using riak, I don't want predis installed.

ivoba replied:
If i remove predis from require-dev the tests will fail:
> 1) Doctrine\Bundle\DoctrineCacheBundle\Tests\Functional\PredisCacheTest::testCacheDriver
> ReflectionException: Class Predis\Client does not exist

If i understand correctly the require-dev block will not be installed when the bundle is part of another composer.json, so i guess when you want to use riak you wont get predis installed unless you have it required in the main composer.json.

xabbuh replied:
Nonetheless, it's a good idea to add it to the `suggest` section too so that you know that you have to install an additional package if you want to use Predis.

ivoba replied:
@xabbuh agreed. i added the suggest section

guilhermeblanco commented:
@ivoba There's only one last thing missing to get this in.
Currently, we fail hard the build if the code does not adhere to our phpcs. Would you mind fix them?

```
FILE: ...DoctrineCacheBundle/DependencyInjection/Definition/PredisDefinition.php
--------------------------------------------------------------------------------
FOUND 4 ERROR(S) AFFECTING 2 LINE(S)
--------------------------------------------------------------------------------
 49 | ERROR | Missing blank line before if statement
 49 | ERROR | Expected "if (...) {\n"; found "if(...){\n"
 57 | ERROR | Missing blank line before if statement
 57 | ERROR | Expected "if (...) {\n"; found "if(...){\n"
--------------------------------------------------------------------------------
```

ivoba commented:
@guilhermeblanco  fixed it.

Should we add: 
   
    <exclude-pattern>*/vendor/*</exclude-pattern>
    <exclude-pattern>*/Tests/*</exclude-pattern>

to the ruleset.xml?  
I ran phpcs locally and had to temporary add it

    vendor/bin/phpcs -pv --standard=ruleset.xml .

mmoreram commented:
Any update on that?

stof commented:
@ivoba they are already excluded from the CLI command on Travis. however, there is still a CS failure in the bundle code. Look at the travis output.

stof commented on Resources/config/schema/doctrine_cache-1.0.xsd:
```diff
@@ -93,6 +94,48 @@
+        <xsd:attribute name="host" type="xsd:string"/>
+        <xsd:attribute name="port" type="xsd:string"/>
+        <xsd:attribute name="database" type="xsd:string"/>
+        <xsd:attribute name="alias" type="xsd:string"/>
```
supporting both attributes and children elements does not make sense IMO (and is totally inconsistent with the Symfony conventions). Btw, it can lead to a WTF behavior if you specify both (I have no idea which one will win)

ivoba replied:
i actually agree, 
i took this over from the other definitions, which all seem to double here.
can anyone explain this?

stof commented on DependencyInjection/Configuration.php:
```diff
@@ -378,6 +379,37 @@ private function addRedisNode()
+            ->children()
+            ->scalarNode('client_id')->defaultNull()->end()
+            ->arrayNode('parameters')
+                ->treatNullLike(array())
```
not needed IMO. It is already done in symfony

stof commented on DependencyInjection/Definition/PredisDefinition.php:
```diff
@@ -0,0 +1,73 @@
+
+        $parameters = $config['parameters'];
+
+        if (is_null($parameters)) {
```
it cannot be null according to the Configuration class

ivoba replied:
aggree, will fix

stof commented on DependencyInjection/Configuration.php:
```diff
@@ -378,6 +379,37 @@ private function addRedisNode()
+            ->arrayNode('parameters')
+                ->treatNullLike(array())
+                ->prototype('scalar')->end()
+                ->defaultValue(array('scheme' => '%doctrine_cache.predis.scheme%',
```
if you expect to have an associative array, you need to use ``->useAttributeAsKey()`` (passing as name the name of the attribute used as key). Otherwise, the lists will not preserve keys when merging multiple files together.

And using a prototyped node is totally inconsistent with your XML XSD, which supports a fixed list of parameters instead.

ivoba replied:
the problem here is that Predis itself doesnt really expect a fixed associative array. 
afaik i will pass the array as is to the connection (with some checks though).
so using useAttributeAsKey would reduce flexibility here.
i would propose to leave it as is and adjust the xsd.

stof replied:
your config currently tells Symfony that you expect a list, so associative keys will be lost when merging config from different files (the fact that keys are not lost when a single file is involved is actually a bug in Symfony).

If you expect the array to be associative (and parameters are, as you want key/value pairs), you must tell it to the config tree

stof commented on Resources/config/services.xml:
```diff
@@ -25,6 +25,11 @@
         <parameter key="doctrine_cache.mongodb.connection.class">MongoClient</parameter>
         <parameter key="doctrine_cache.mongodb.server">localhost:27017</parameter>
+        <parameter key="doctrine_cache.predis.class">Doctrine\Common\Cache\PredisCache</parameter>
+        <parameter key="doctrine_cache.predis.client.class">Predis\Client</parameter>
```
defining parameters for service classes is discouraged in Symfony since 2.6 (and we will remove them for all core services in Symfony 3.0). so IMO, you should not add new service class parameters here either (existing ones should be kept for BC reasons)

ivoba replied:
i would rather keep it for now, as the whole bundle uses this, just to keep it consistent

stof replied:
Well, I think that following the way of Symfony 2.6 (i.e. not introducing them for new services and considering them as a deprecated extension point for existing services so that people should stop overwriting such parameters) is a better idea than continuing to introduce more parameters, which don't bring much value (they are generally not a good way to replace a service as a different class will generally have different deps too) and consume memory at runtime (a typical Symfony project currently has hundreds or even thousands such parameters never used at runtime but loaded in memory, which was an argument to eliminate the parameters which don't bring value)

ivoba replied:
ok convinced, will fix

stof commented on composer.json:
```diff
@@ -49,7 +49,11 @@
+        "predis/predis":                         "~0.8"
+    },
+    "suggest": {
+       "predis/predis":                          "Required if you want to use Predis as cache backend"
```
this should not be in the suggestions here IMO. ``doctrine/cache`` is the way where suggestions for cache dependencies should be done IMO

ivoba replied:
agree

stof commented on DependencyInjection/Configuration.php:
```diff
@@ -378,6 +379,37 @@ private function addRedisNode()
+                                     'host' => '%doctrine_cache.predis.host%',
+                                     'port' => '%doctrine_cache.predis.port%'))
+            ->end()
+            ->arrayNode('options')
```
same issues here than for the parameters

stof commented on DependencyInjection/Configuration.php:
```diff
@@ -378,6 +379,37 @@ private function addRedisNode()
+                ->prototype('scalar')->end()
+                ->defaultValue(array('scheme' => '%doctrine_cache.predis.scheme%',
+                                     'host' => '%doctrine_cache.predis.host%',
+                                     'port' => '%doctrine_cache.predis.port%'))
```
This does not ensure that these keys will be present., as the default value of a prototyped node is used only when the node is omitted entirely. It is not merged with the configured values

ivoba commented:
i updated the PR, the CS issue is fixed and i adjusted most things mentioned by @stof.
pls review.
the failing test is not part of the the PR but was there already.

mmoreram commented:
Ping!

C'mon! Let's merge this PR if is ready, because can reduce a lot the dependency set of many projects :)

rodrigodiez commented:
+1 Waiting for predis support so we can migrate legacy apc dependant code

rodrigodiez commented:
Ping! Any news on this?

ivoba commented:
@guilhermeblanco can you have look pls and merge this, when all is fine.

rodrigodiez commented:
Hi guys,

Some feedback would be appreciated. It the PR is not ready to be merged then please spot the problems so they can be fixed.

If the responsible for this call is on holidays (it seems so) please take over this PR and assign it to who can make this to flow

Thanks in advance

mremi commented:
+1 would be great to merge it

ivoba commented:
Hey guys, i merged the latest master into the PR.
Tests are passing now.

guilhermeblanco commented:
@ivoba why parameters are optional? It looks like the parameters should be top-level attribute definitions, not part of a generic array.

ivoba commented:
@guilhermeblanco 
I guess i left the parameters undefined and optional, because
1. Predis accepts an array of config or a URI string, so i dont want to have this restricted by the bundle
2. i wanted to keep parameters open to additional paramters like cluster config as array, alias etc. to describe this in schema would be rather cumbersome
3. Predis works also without parameters "without passing any connection parameter, Predis assumes 127.0.0.1 and 6379"

phoenixgao commented:
Will this be merged?

need it :P

lasektom commented:
Can this be merge soon? Please!

stof commented on DependencyInjection/Configuration.php:
```diff
@@ -379,6 +380,35 @@ private function addRedisNode()
+            ->children()
+            ->scalarNode('client_id')->defaultNull()->end()
+            ->arrayNode('parameters')
+                ->prototype('scalar')->end()
```
my comment about the need to mark the array as an associative array (which got hidden by github due to other changes) is still true. the current config does not allow the user to set keys in the config (they work in the simple case only because of a bug, but would break when merging config from multiple files)

ivoba replied:
@stof how can i define "parameters" to be whatever comes?
Predis can work with a string:
`parameters: 'tcp://10.0.0.1:6379'`

with array:
parameters:
`    [
    'scheme' => 'tcp',
    'host'   => '10.0.0.1',
    'port'   => 6379,
    ]`
with nested arrays for cluster configuration:
`[
    'tcp://10.0.0.1?alias=first-node',
    ['host' => '10.0.0.2', 'alias' => 'second-node'],
]`
even with null, it will fallback to the usual redis parameters
`parameters: null `

ivoba replied:
i guess it would be:

    ->variableNode('parameters')->defaultNull()->end()

stof commented on DependencyInjection/Configuration.php:
```diff
@@ -379,6 +380,35 @@ private function addRedisNode()
+                ->prototype('scalar')->end()
+                ->defaultValue(array('scheme' => '%doctrine_cache.predis.scheme%',
+                                     'host' => '%doctrine_cache.predis.host%',
+                                     'port' => '%doctrine_cache.predis.port%'))
```
if the user set a host only in their config file, the scheme and port will not be configured at all anymore

stof commented on README.md:
```diff
@@ -405,6 +405,9 @@ $aclCache = $this->container->get('security.acl.cache');
     - timeout       - redis connection timeout
     - database      - redis database selection (integer)
+#### predis
+    - parameters - array of parameters: {'schema': 'tcp', 'host': '10.0.0.1', 'port': '6379'}
```
your code uses ``scheme``, not ``schema``

stof commented on Resources/config/services.xml:
```diff
@@ -25,6 +25,10 @@
         <parameter key="doctrine_cache.mongodb.collection.class">MongoCollection</parameter>
         <parameter key="doctrine_cache.mongodb.connection.class">MongoClient</parameter>
         <parameter key="doctrine_cache.mongodb.server">localhost:27017</parameter>
+        <parameter key="doctrine_cache.predis.client.class">Predis\Client</parameter>
```
is this parameter actually useful ?

ivoba replied:
i think it is, because its used in the BundleExtension, the alternative would be to hardcode the class in PredisDefinition, which is probably not better

stof replied:
well, defining service classes as parameters is not a recommended practice anymore (any service added in Symfony after 2.6 does not have it, and 3.0 changed all older services)

guilhermeblanco replied:
There's usage for this though... If you enhanced the default API, this is a perfect solution instead of writing a `CompilerPass`. Also, the entire bundle is coded this way, I see no harm following the standards of the bundle

stof commented on Tests/Functional/Fixtures/config/predis.xml:
```diff
@@ -0,0 +1,24 @@
+                    <host>127.0.0.1</host>
+                    <port>6379</port>
+                    <database>2</database>
+                    <timeout>10</timeout>
```
this does not look at all like the way Symfony configs for prototyped array nodes look like in XML, meaning it is incompatible with any existing tooling (IDE integrations, etc...).

And the XSD schema does not allow to provide any autocompletion either btw.

guilhermeblanco commented:
I noted to creator to move the parameters as specific configuration flags. A soon as it gets ready, merge will happen or it'll happen once I do another pass over the project to address bugs.

ivoba commented:
ok guys,  i adjusted the parameters definition.

i only added basic parameters for tcp scheme.  
advanced and unusual config must use their own client service via client_id config.

guilhermeblanco commented on DependencyInjection/Configuration.php:
```diff
@@ -379,6 +380,44 @@ private function addRedisNode()
+            ->end()
+        ;
+
+//        profile: which profile to use in order to match a specific version of Redis.
```
Remove this comment lines and we're good to go! =)

ivoba replied:
aaaargh, removed
